### PR TITLE
Istio - intial commit

### PR DESCRIPTION
**Pull request by srinivasa-vasu**
#7337 
- tested with both eureka and consul
- included routerule and destinationpolicy rules
This is with minimal change to the existing code base. We can slowly introduce more features to it and over the period of time redundant components can be decoupled.

### CODE REVIEW

File: generators/kubernetes/templates/deployment.yml.ejs [hunk 1]
```diff
@@ -76,12 +81,26 @@ spec:
...
         - name: SPRING_CLOUD_CONSUL_PORT
           value: "8500"
+          <%_ if (app.istioSupportOpt !== 'no') { _%>
+        - name: SPRING_CLOUD_CONSUL_DISCOVERY_PREFER_IP_ADDRESS
+          value: "false"
+        - name: SPRING_CLOUD_CONSUL_DISCOVERY_HOSTNAME
+          value: <%= app.baseName.toLowerCase() %>
+        - name: SPRING_CLOUD_CONSUL_DISCOVERY_SERVICE_NAME
+          value: <%= app.baseName.toLowerCase() %>
+          <%_ } _%>
```
Comment (saturnism) [inline generators/kubernetes/templates/deployment.yml.ejs]: can we add one more conditional env var to propagate additional Istio headers? e.g.:

```
- name: SPRING_SLEUTH_PROPAGATION_KEYS
   value: "x-request-id,x-ot-span-context"
```

Comment (srinivasa-vasu) [reply generators/kubernetes/templates/deployment.yml.ejs]: Sure, I guess sleuth is not available as a dependency. Have to add it to the build file first.

Comment (PierreBesson) [reply generators/kubernetes/templates/deployment.yml.ejs]: Sleuth is available when packaging the app with the zipkin maven profile.

File: generators/kubernetes/prompts.js [hunk 2]
```diff
@@ -102,3 +104,64 @@ function askForIngressDomain() {
...
+    if (this.composeApplicationType === 'monolith') {
+        this.istioSupportOpt = 'no';
+        return;
+    }
+    const done = this.async();
+
+    const prompts = [{
+        type: 'list',
+        name: 'istioSupportOpt',
+        message: 'Choose your preference for Istio Service Mesh',
```
Comment (PierreBesson) [inline generators/kubernetes/prompts.js]: I would use the following:
- name: `istio`
- message: `Do you want to setup Istio ?`
- value: 'manualInjection`or value: `automaticInjection`

File: generators/kubernetes/prompts.js [hunk 3]
```diff
@@ -102,3 +104,64 @@ function askForIngressDomain() {
...
+
+function askForIstioRouteFiles() {
+    if (this.istioSupportOpt === 'no') return;
+    const done = this.async();
+
+    const prompts = [{
+        type: 'list',
+        name: 'istioRouteObjs',
+        message: 'Do you want to generate Istio route files?',
+        choices: [
```
Comment (PierreBesson) [inline generators/kubernetes/prompts.js]: name: `istioRoute`
value: boolean

File: generators/kubernetes/templates/deployment.yml.ejs [hunk 4]
```diff
@@ -16,17 +16,22 @@
  See the License for the specific language governing permissions and
  limitations under the License.
 -%>
-apiVersion: extensions/v1beta1
+apiVersion: apps/v1
```
Comment (PierreBesson) [inline generators/kubernetes/templates/deployment.yml.ejs]: 👍

File: generators/kubernetes/templates/deployment.yml.ejs (hunk 1 above)
Comment (srinivasa-vasu) [reply generators/kubernetes/templates/deployment.yml.ejs]: @saturnism done!

File: generators/kubernetes/prompts.js (hunk 3 above)
Comment (srinivasa-vasu) [reply generators/kubernetes/prompts.js]: done

File: generators/kubernetes/templates/deployment.yml.ejs (hunk 1 above)
Comment (saturnism) [reply generators/kubernetes/templates/deployment.yml.ejs]: nice!

**Comment by saturnism** [comment]
i think this is a good first step! i'll add additional checklist items that we can address later.

**Comment by deepu105** [comment]
builds failed due to some network/yarn issue. I restarted them

**Comment by saturnism** [comment]
@srinivasa-vasu one last thing, if istio support is enabled, then if when generating ingress, it should use istio ingress class w/ annotation:

```
  kubernetes.io/ingress.class: istio
```

**Comment by saturnism** [comment]
#7695 is also ready for review, but will need all the changes made in this PR for it to work properly.

**Comment by PierreBesson** [comment]
@srinivasa-vasu For the options you added as manual/auto sidecar injection, as this only impact the apply.sh script and it's potentially subject to change in the lifetime of the project so I'm thinking that maybe it could be a script parameter like this: `./apply.sh --sidecar-injection=manual/auto` and if none was given then output a message to the user.

**Comment by saturnism** [comment]
can we merge the MVP first and add enhancements via https://github.com/jhipster/generator-jhipster/issues/7708 ?   Since not everyone will be using apply script either.

I feel we can track it separately, where it actually generates the YAML w/ `istio kube-inject` on the yaml file directly. The YAML with the injection is the final output.

**Comment by PierreBesson** [comment]
I agree let's merge this quickly with a BETA tag. It's better to release early but I don't want to add useless questions especially because we have to provide configuration fallbacks when changing questions for people that regenerate from the same yo-rc.

**Comment by saturnism** [comment]
Ah I see! Fair point because the answer is stored.

**Comment by jdubois** [comment]
I just restarted the Travis build, I hope it will pass - yes let's merge this as soon as possible

**Comment by saturnism** [comment]
exciting! :D

**Pull request merged**